clang/tools/cir-translate/cir-translate.cpp

```diff
@@ -30,6 +38,103 @@ extern std::unique_ptr<llvm::Module> lowerDirectlyFromCIRToLLVMIR(
     bool disableVerifier = false, bool disableCCLowering = false,
     bool disableDebugInfo = false);
 } // namespace direct
+
+namespace {
+
+std::string determineDataLayoutByTriple(llvm::StringRef rawTriple) {
+  llvm::Triple triple(rawTriple);
+  // Data layout is fully determined by the target triple. Here we only pass the
+  // triple to get the data layout.
+  clang::TargetOptions targetOptions;
+  targetOptions.Triple = rawTriple;
+  // FIXME: AllocateTarget is a big deal. Better make it a global state.
+  auto targetInfo =
+      clang::targets::AllocateTarget(llvm::Triple(rawTriple), targetOptions);
+  if (!targetInfo) {
+    llvm::errs() << "error: invalid target triple '" << rawTriple << "'\n";
+    exit(1);
+  }
+  return targetInfo->getDataLayoutString();
+}
+
+/// The goal of this option is to ensure that the triple and data layout specs
+/// are always available in the ClangIR module. With this requirement met, the
+/// behavior of this option is designed to be as intuitive as possible, as shown
+/// in the table below:
+///
+/// +--------+--------+-------------+-----------------+-----------------------+
+/// | Option | Triple | Data Layout | Behavior Triple | Behavior Data Layout  |
+/// +========+========+=============+=================+=======================+
+/// | T      | T      | T           | Overwrite       | Derive from triple    |
+/// | T      | T      | F           | Overwrite       | Derive from triple    |
+/// | T      | F      | T           | Overwrite       | Derive from triple    |
+/// | T      | F      | F           | Overwrite       | Derive from triple    |
+/// | F      | T      | T           |                 |                       |
+/// | F      | T      | F           |                 | Derive from triple    |
+/// | F      | F      | T           | Set default     | Derive from triple    |
+/// | F      | F      | F           | Set default     | Derive from triple    |
+/// +--------+--------+-------------+-----------------+-----------------------+
+llvm::cl::opt<std::string>
+    targetTripleOption("target",
+                       llvm::cl::desc("Specify a default target triple when "
+                                      "it's not available in the module"),
+                       llvm::cl::init(""));
+
+std::string prepareCIRModuleTriple(mlir::ModuleOp mod) {
+  std::string triple = targetTripleOption;
+
+  // Treat "" as the default target machine.
+  if (triple.empty()) {
+    // FIXME: Use `llvm::sys::getDefaultTargetTriple()` in the future. Currently
+    // ClangIR does not support OS like Windows, which may lead to CI failure.
```

Comment:
I'm not sure `llvm::sys::getDefaultTargetTriple()` is what we want though, relying on the host triple might not provide much better results either. I suggest just to mention in the comments that we support a couple of targets so far and not providing a triple will default to `x86_64-unknown-linux-gnu`.

Btw, is it possible to emit a warning here (not sure a diagnostic engine is a easy setup at this code path)? If so we should probably emit a warning saying `warning: no target triple provided, assuming x86_64-unknown-linux-gnu`.